framework/source/class/qx/core/Id.js

```diff
@@ -0,0 +1,166 @@
+
+************************************************************************ */
+
+/**
+ * Provides a registry of top level objects
+ */
+qx.Class.define("qx.core.Id", {
+  extend: qx.core.Object,
+  include: [ qx.core.MObjectId ],
+  type: "singleton",
+  
+  members: {
+    
+    __registeredObjects: null,
+    __registeredIdHashes: null,
+
+    /*
+     * @Override
+     */
+    _createObjectImpl: function(id) {
+      if (this.__registeredObjects) {
+        var obj = this.__registeredObjects[id];
+        if (obj !== undefined) {
+          return obj;
+        }
+      }
+      
+      switch(id) {
+        case "application":
+          return qx.core.Init.getApplication() || undefined;
+      }
+      
+      return; // undefined
+    },
+    
+    /**
+     * Returns an object path which can be used to locate an object anywhere in the application
+     * with a call to `qx.core.Id.getObject()`.
+     * 
+     * This will return null if it is not possible to calculate a path because one of the
+     * ancestors has a null `objectId`.
+     * 
+     * This will also return null if the top-most ancestor is not one of the globals registered
+     * with `registerObject` or a known global (such as the application); however, by passing
+     * `false` as the `mustBeRegistered` parameter, this restriction is removed and a path will
+     * still be returned (it's just that in this case, you will have to know the the root object
+     * to query and `qx.core.Id.getObject` will not work on that path)  
+     * 
+     * @param obj {qx.core.Object} the object
+     * @param mustBeRegistered {Boolean} whether the
+     * @return {String} full path to the object
+     */
+    getAbsoluteIdOf: function(obj, mustBeRegistered) {
+      var segs = [];
+      var application = qx.core.Init.getApplication();
+      while (obj) {
+        var id = obj.getObjectId();
+        if (!id) {
+          this.error("Cannot determine an absolute Object ID because one of the ancestor ObjectID's is null");
+          return null;
+        }
+        segs.unshift(id);
+        var owner = obj.getOwner();
+        if (owner) {
+          // Find the ID of the owner, *if* it is registered as a top level object
+          var ownerId = null;
+          if (owner === application) {
+            ownerId = "application";
+          } else {
+            ownerId = this.__registeredIdHashes && this.__registeredIdHashes[owner.toHashCode()] || null;
+          }
+
+          // When we have found the ID of a top level object, add it to the path and stop 
+          if (ownerId) {
+            segs.unshift(ownerId);
+            break;
+          }
+        } else {
+          if (mustBeRegistered === false) {
+            break;
+          }
+          this.error("Cannot determine a global absolute Object ID because the topmost object is not registered");
+          return null;
+        }
+        obj = owner;
+      }
+      var path = segs = segs.join("/");
```

Comment:
`segs` is a local variable. It's not used after this line, in this method, so need not be assigned to.